File: Plugins/Orimath.Basics.View/ViewModels/WorkspaceViewModel.cs
using Orimath.Controls;
using Orimath.Plugins;

namespace Orimath.Basics.View.ViewModels
{
    public class WorkspaceViewModel : NotifyPropertyChanged
    {
        private readonly IWorkspace _workspace;
        private readonly IViewPointConverter _pointConverter;
        private readonly IDispatcher _dispatcher;

        public PaperViewModel Paper { get; }

        public WorkspaceViewModel(IWorkspace workspace, IViewPointConverter pointConverter, IDispatcher dispatcher)
        {
            _workspace = workspace;
            _pointConverter = pointConverter;
            _dispatcher = dispatcher;

            Paper = new PaperViewModel(_workspace.Paper, _pointConverter, _dispatcher);
        }

        private OperationTarget ToModelTarget(ScreenOperationTarget target)
        {
            return new OperationTarget(_pointConverter.ViewToModel(target.Point), target.Layer, target.Target);
        }

        public void OnClick(ScreenOperationTarget target, OperationModifier modifier)
        {
            if (_workspace.CurrentTool.Value is IClickTool tool)
                _dispatcher.OnBackgroundAsync(() => tool.OnClick(ToModelTarget(target), modifier));
        }

        public bool BeginDrag(ScreenOperationTarget source, OperationModifier modifier) =>
             (_workspace.CurrentTool.Value as IDragTool)?.BeginDrag(ToModelTarget(source), modifier) ?? false;

        public bool DragEnter(ScreenOperationTarget source, ScreenOperationTarget target, OperationModifier modifier) =>
            (_workspace.CurrentTool.Value as IDragTool)?.DragEnter(ToModelTarget(source), ToModelTarget(target), modifier) ?? false;

        public void DragLeave(ScreenOperationTarget source, ScreenOperationTarget target, OperationModifier modifier) =>
            (_workspace.CurrentTool.Value as IDragTool)?.DragLeave(ToModelTarget(source), ToModelTarget(target), modifier);

        public void DragOver(ScreenOperationTarget source, ScreenOperationTarget target, OperationModifier modifier) =>
            (_workspace.CurrentTool.Value as IDragTool)?.DragOver(ToModelTarget(source), ToModelTarget(target), modifier);

        public void Drop(ScreenOperationTarget source, ScreenOperationTarget target, OperationModifier modifier)
        {
            if (_workspace.CurrentTool.Value is IDragTool tool)
                _dispatcher.OnBackgroundAsync(() => tool.Drop(ToModelTarget(source), ToModelTarget(target), modifier));
        }
    }

    public class ScreenOperationTarget
    {
        public System.Windows.Point Point { get; }
        public ILayerModel Layer { get; }
        public DisplayTarget Target { get; }

        public ScreenOperationTarget(System.Windows.Point point, (ILayerModel, DisplayTarget) target)
        {
            Point = point;
            (Layer, Target) = target;
        }
    }
}

File: Plugins/Orimath.Basics.View/ViewModels/WorkspaceViewModel.fs
namespace Orimath.Basics.View.ViewModels
open Orimath.Core
open Orimath.Controls
open Orimath.Plugins

type WorkspaceViewModel(workspace: IWorkspace, pointConverter: IViewPointConverter, dispatcher: IDispatcher) =
    inherit NotifyPropertyChanged()
    let toModelTarget (target: ScreenOperationTarget) =
        let point = pointConverter.ViewToModel(target.Point)
        let adjustedPoint =
            match target.Target with
            | DisplayTarget.Point(point) -> point
            | DisplayTarget.Crease(crease) -> Line.perpFoot point crease.Line
            | DisplayTarget.Edge(edge) -> Line.perpFoot point edge.Line
            | DisplayTarget.Layer _ -> point
        { Point = adjustedPoint
          Layer = target.Layer
          Target = target.Target }

    member val Paper = PaperViewModel(workspace.Paper, pointConverter, dispatcher)

    member _.OnClick(target, modifier) =
        match workspace.CurrentTool.Value with
        | :? IClickTool as tool ->
            dispatcher.Background { tool.OnClick(toModelTarget(target), modifier) }
        | _ -> ()

    member _.BeginDrag(source, modifier) =
        match workspace.CurrentTool.Value with
        | :? IDragTool as tool -> tool.BeginDrag(toModelTarget(source), modifier)
        | _ -> false

    member _.DragEnter(source, target, modifier) =
        match workspace.CurrentTool.Value with
        | :? IDragTool as tool ->
            tool.DragEnter(toModelTarget(source), toModelTarget(target), modifier)
        | _ -> false

    member _.DragLeave(source, target, modifier) =
        match workspace.CurrentTool.Value with
        | :? IDragTool as tool ->
            tool.DragLeave(toModelTarget(source), toModelTarget(target), modifier)
        | _ -> false

    member _.DragOver(source, target, modifier) =
        match workspace.CurrentTool.Value with
        | :? IDragTool as tool -> tool.DragOver(toModelTarget(source), toModelTarget(target), modifier)
        | _ -> false

    member _.Drop(source, target, modifier) =
        match workspace.CurrentTool.Value with
        | :? IDragTool as tool ->
            dispatcher.Background { tool.Drop(toModelTarget(source), toModelTarget(target), modifier) }
        | _ -> ()

    member _.CancelDrag(source, modifier) =
        match workspace.CurrentTool.Value with
        | :? IDragTool as tool ->
            dispatcher.Background { tool.CancelDrag(toModelTarget(source), modifier) }
        | _ -> ()
